[Backend] Fix #3256: Add submission meta attributes in download_all_submissions API

afif-fahim opened the pull request:
fix #3256

afif-fahim commented:
@Ram81 can you review this PR, please?

Ram81 reviewed (approved):
LGTM

Ram81 commented on apps/challenges/views.py:
```diff
@@ -1840,6 +1841,7 @@ def download_all_submissions(
                             submission["created_at"],
                             submission["submission_result_file"],
                             submission["submission_metadata_file"],
+                            submission["submission_metadata"],
```
@afif-fahim can we just fetch fields `name` and `value` for each attribute in submission metadata and just send that in this column

afif-fahim replied:
@Ram81, can you review again please?
After changes:
![image](https://user-images.githubusercontent.com/33936462/109487660-b6850080-7aae-11eb-9ac3-c30e7a8c1c07.png)

Ram81 commented on apps/challenges/views.py:
```diff
@@ -1840,6 +1843,12 @@ def download_all_submissions(
                             submission["created_at"],
                             submission["submission_result_file"],
                             submission["submission_metadata_file"],
+                            ",".join(
+                                f'{key}: {value}'
+                                for key, value in submission[
+                                    "submission_metadata"
+                                ].items()
+                            ),
```
@afif-fahim can we move this to line 1813.

afif-fahim replied:
@Ram81, do you mean before `writer.writerow`?

Ram81 reviewed (changes requested):
@afif-fahim we also want to show the option to exclude or include submission meta attributes in the file on UI. Can you please add the changes on UI side too.

afif-fahim commented:
@Ram81 sure! I will try.

Ram81 commented:
@afif-fahim any updates?

afif-fahim commented:
@Ram81, Sorry I didn't have the chance to work on this! I will try tonight.

afif-fahim commented:
@Ram81, Added the changes on UI and updated the PR.

![image](https://user-images.githubusercontent.com/33936462/110375182-d3f03680-807b-11eb-9405-6c59f8c73767.png)

Ram81 reviewed (changes requested):
@afif-fahim I tried pulling the changes locally and downloading submissions with some meta attributes and it seems to be throwing Http status 500. Here's the error for your reference

```
AttributeError at /api/challenges/21/phase/41/download_all_submissions/csv/
'list' object has no attribute 'items'
```
Can you please make the required changes

afif-fahim commented:
@Ram81, Can submission meta attributes consist of multiple dictionaries?

afif-fahim commented:
@Ram81 ping.

Ram81 commented:
@afif-fahim you can refer the challenge config with meta attributes from [here](https://github.com/Cloud-CV/EvalAI-Starters/blob/master/challenge_config.yaml#L49)

afif-fahim commented:
@Ram81, can you review this please. If this is the right way to fetch the attributes of submission metadata?

![image](https://user-images.githubusercontent.com/33936462/110961074-9a6c4380-8379-11eb-8bc1-3b920cb2ec89.png)

Ram81 commented on apps/challenges/views.py:
```diff
@@ -1807,6 +1808,13 @@ def download_all_submissions(
+                    submission["submission_metadata"] = ",".join(
+                        f'{key}: {value}'
+                        for metadata in submission["submission_metadata"]
+                        for key, value in metadata.items()
```
@afif-fahim Sorry for the delay. We only need to fetch field name and field value. Current changes are giving all metadata sent by frontend in this field. This would be difficult for hosts to interpret. Here's what the current output looks like 

```
"name: TextAttribute,type: text,value: None,required: False,$$hashKey: object:201,description: Sample,name: SingleOptionAttribute,type: radio,value: B,options: ['A', 'B', 'C'],$$hashKey: object:202,description: Sample,name: MultipleChoiceAttribute,type: checkbox,values: ['gamma'],options: ['alpha', 'beta', 'gamma'],$$hashKey: object:203,description: Sample,name: TrueFalseField,type: boolean,value: False,required: True,$$hashKey: object:204,description: Sample"
```

It should be:
```
TextAttribute: None, SingleOptionAttribute: B

```

where `TextAttribute` is the name field in each metadata entry and value for each field name should be value from the stored object. Let me know if you have any doubts
```

afif-fahim replied:
@Ram81, Sorry for the misunderstanding! I have pushed the new changes.
